pkg/app/mutate_cluster_app.go

```diff
@@ -0,0 +1,159 @@
+
+	"github.com/giantswarm/app-admission-controller/pkg/mutator"
+)
+
+type providerReleaseConfig struct {
+	lastClusterAppVersionWithoutRelease *semver.Version
+}
+
+var providersReleaseConfig = map[string]providerReleaseConfig{
+	"cluster-aws": {
+		lastClusterAppVersionWithoutRelease: semver.New(0, 76, 1, "", ""),
+	},
+}
+
+func (m *Mutator) mutateClusterApp(ctx context.Context, app v1alpha1.App) ([]mutator.PatchOperation, error) {
+	m.logger.Debugf(ctx, "Cluster app mutation for setting App version based on the release. App/Cluster:%s, Namespace:%s\n",
+		app.Name, app.Namespace)
+
+	releaseConfig, providerSupportsReleases := providersReleaseConfig[app.Spec.Name]
+	if !providerSupportsReleases {
+		return nil, nil
+	}
+	lastClusterAppVersionWithoutRelease := releaseConfig.lastClusterAppVersionWithoutRelease
+
+	// Check if app is a cluster-$provider app
+	isClusterApp := (app.Spec.Catalog == "cluster" || app.Spec.Catalog == "cluster-test") && strings.HasPrefix(app.Spec.Name, "cluster-")
+	if !isClusterApp {
+		return nil, nil
+	}
+
+	// Check if the App.spec.version is already set to a cluster-aws release from which we upgrade to new releases
+	if app.Spec.Version != "" {
+		currentClusterAppVersion, err := semver.NewVersion(app.Spec.Version)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+		if !currentClusterAppVersion.GreaterThan(lastClusterAppVersionWithoutRelease) {
+			// This is a cluster that is using an older version of the cluster-$provider app, a version that does
+			// not use new releases, so it does not have release version. Therefore, we don't overwrite cluster app
+			// version here.
+			//
+			// True for cluster-aws versions older or equal to v0.76.1.
+			return nil, nil
+		}
+
+		// Additional temporary check during testing.
+		//
+		// Check if this is a dev build on top of the latest version.
+		isNewDevBuild := currentClusterAppVersion.Major() == lastClusterAppVersionWithoutRelease.Major() &&
+			currentClusterAppVersion.Minor() == lastClusterAppVersionWithoutRelease.Minor() &&
+			currentClusterAppVersion.Patch() == lastClusterAppVersionWithoutRelease.Patch() &&
+			currentClusterAppVersion.Prerelease() != ""
+		// The last "standalone" cluster-aws version will be v0.76.1, but current dev builds on top of it are per semver
+		// actually older, and here we want to treat custom dev build with Releases as newer.
+		clusterAppReleasesDevBuild := semver.New(0, 76, 1, "b76af2c26f4224ffb0d718e940e232fac05c89a0", "")
+		isClusterAppReleasesDevBuild := currentClusterAppVersion.Equal(clusterAppReleasesDevBuild)
+
+		if isNewDevBuild && !isClusterAppReleasesDevBuild {
+			// This is a development cluster that is using a dev version of the cluster-$provider app that does not use
+			// new releases, so it does not have release version. Therefore, we don't overwrite cluster app
+			// version here.
+			return nil, nil
+		}
+	}
+
+	userConfigMapName := app.Spec.UserConfig.ConfigMap.Name
```

Comment:
if I understand correctly, the `global.release.version` is being retrieved only from the UseConfig configmap. That's against the App specification, where values are merged from different sources.
Yes, it's unlikely that we'll see that use-case for cluster apps; but still I think the same rule should apply.

maybe you could make use of https://github.com/giantswarm/app/blob/main/pkg/values/configmap.go#L23

Reply:
> if I understand correctly, the global.release.version is being retrieved only from the UseConfig configmap.

Yes, it is a user-provided value that should really never be set from any other place.

`global.release.version` represents a workload cluster release version for the new CAPI releases.

I cannot imagine that we will ever set this in some global place like catalog values and cluster values, because that would enforce the same release version .

And in general, for `cluster-<provider>` apps only user values are relevant, other config layers are not used (or there are maybe few leftover use-cases that we should remove). Injecting Helm values from other places (outside of user values, so e.g. catalog values, cluster values, etc) is an issue for cluster apps and has caused problems already, and this is a consequence of the choice to use App platform for deploying clusters (which was IMO a wrong choice from the start, because cluster "apps" are not really apps and have different needs and requirements compared to "normal" apps, so hopefully at some point we move away from App platform for deploying CAPI/CAPx and other cluster resources). See my comment [here](https://github.com/giantswarm/roadmap/issues/3187#issuecomment-1931888908) for more details about this.

Reply:
If we reach the point where we have to bend `cluster-<provider>` apps behaviour to App platform rules that do not make sense for deploying workload clusters, IMO that is the point when we should stop using App platform for deploying workload clusters and we should switch to Flux's HelmRelease (or some other way which would make sense).

Reply:
> Yes, it is a user-provided value that should really never be set from any other place.

The user has the freedom to provider user values in different ways. That's a feature of the app platform (userConfig, extraConfigs).

> global.release.version represents a workload cluster release version for the new CAPI releases. This value is required by the cluster chart.

that's clear.

> I cannot imagine that we will ever set this in some global place like catalog values and cluster values, because that would enforce the same release version across many clusters. Yeah, you can override it, but if some other config layers sets this then cluster chart will not complain because it got the value, and that may cause all sorts of issues and we cannot know if the customer did not specify the value or if they specified it in some incorrect way that we didn't catch.

I understand and agree that is not desirable to set this value in a global scope as the catalogs are.

> And in general, for cluster-<provider> apps only user values are relevant, other config layers are not used (or there are maybe few leftover use-cases that we should remove). 

User values can come from extraConfig as well. Aside that, I wouldn't close the door to that, just because it's an unknown at this stage. I'd argue that once CAPI is under extensive use by customer we might start to see the need for that.

> Injecting Helm values from other places (outside of user values, so e.g. catalog values, cluster values, etc) is an issue for cluster apps and has caused problems already, and this is a consequence of the choice to use App platform for deploying clusters (which was IMO a wrong choice from the start, because cluster "apps" are not really apps and have different needs and requirements compared to "normal" apps, so hopefully at some point we move away from App platform for deploying CAPI/CAPx and other cluster resources). See my comment https://github.com/giantswarm/roadmap/issues/3187#issuecomment-1931888908 for more details about this.

I can't agree more with this. I also think that Apps is not the right tool for clusters; starting with the semantics but also that a cluster needs some dedicated interface. But anyways, here we are.

In this context, I still don't think that breaking the App platform behavior for one "kind" of app (which is based only on the name  & catalog) is really a good idea.
Imagine the case where someone decides to actually do merging and set the value in the `userConfig` and in `extraConfigs` (by mistake or on purpose, it doesn't matter). Now this cluster will have the `cluster-<provider>` version from the release in `userConfig`; but the helm lookups will be done using the release from the `extraConfigs`. This is an effect of `app-admission-controller` and `app-operator` not behaving the same. 

This is one example trying to picture why I think we should not have this kind of exceptions. People (and documentation) expect this behavior, leading to weird cases and not so easy to debug behaviors.

Reply:
by the way, I think consistency is important to avoid possible problems in the future and the solution is not that expensive given the fact that there is already a library can be used.

Having said that, if the owner of the app platform is happy with this, I'm ok. definitively not a blocker from my side.

Reply:
> User values can come from extraConfig as well.

Right, I forgot about extraConfigs. Today in CAPI, for workload clusters, we just use user config, as that is rendered by our tooling (kubectl-gs and happa). We use user config for our clusters, and so far our customers as well. We and customers do use extraConfig, but for regular apps.

That being said, extraConfig can be a legit case here, so I'll add what is needed for extraConfigs (maybe just not immediately in this PR as we need this change ASAP and the extraConfig case here, although possible, is still non-existent in practice).

Reply:
Thanks for review here btw!

Reply:
> Imagine the case where someone decides to actually do merging and set the value in the userConfig and in extraConfigs (by mistake or on purpose, it doesn't matter).

![](https://i.giphy.com/media/v1.Y2lkPTc5MGI3NjExMHNtdTM3eHIzNWpjZmt5YTBxbmdwZG1raW9ubmdmd2x3eHUyOWN3ZSZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/nZjB4cAWkYRcA/giphy.gif)